hyperbolic-rs/preprocess.py

```diff
@@ -21,38 +21,114 @@ def movielens_to_dict(dataset_path):
     for uid in samples:
         sorted_items = sorted(samples[uid], key=lambda p: p[1])
-        samples[uid] = np.array([pair[0] for pair in sorted_items]).astype('int64')
-    return samples
+        sorted_samples[uid] = [pair[0] for pair in sorted_items]
+    return sorted_samples
 
 
-def create_splits(samples, random=False):
+def build_movieid2title(dataset_path):
+    filename = "movies.dat"
+    movieid2title = {}
+    with open(Path(dataset_path) / filename, "r", encoding="ISO-8859-1") as f:
+        for line in f:
+            line = line.strip("\n").split("::")
+            movieid2title[line[0]] = line[1]
+    return movieid2title
+
+
+def csv_to_dict(file_path, min_interactions=10):
+    """
+    Maps raw csv interactions file to a Dictonary.
+
+    :param dataset_path: Path to file containing interactions in a format
+        user_id,keen_id
+    :param min_interactions: users with less than min_interactions are discarded
+    :return: Dictionary containing users as keys, and a numpy array of items the user
+      interacted with, sorted by the time of interaction.
+    """
+    samples = {}
+    with open(file_path, 'r') as f:
+        next(f)
+        for line in f:
+            line = line.strip('\n').split(',')
+            uid = line[0]
+            iid = line[1]
+            if uid in samples:
+                samples[uid].add(iid)
+            else:
+                samples[uid] = {iid}
+
+    all_items = set()
+    for ints in samples.values(): all_items.update(ints)
+    interactions = sum([len(v) for v in samples.values()])
+    print(f"Processed users: {len(samples)}. Processed items: {len(all_items)}. Interactions: {interactions}")
+
+    filtered_samples = {uid: ints for uid, ints in samples.items() if len(ints) >= min_interactions}
+
+    filtered_items = set()
+    for ints in filtered_samples.values(): filtered_items.update(ints)
+    interactions = sum([len(v) for v in filtered_samples.values()])
+    print(f"After filtering: users: {len(filtered_samples)}, items: {len(filtered_items)}, interactions: {interactions}")
+
+    return filtered_samples
+
+
+def build_iid2title(item_id_key, item_title_key):
+    keen_metadata = "data/keen/keens_and_gems_25_june_2020.json"
+    iid2title = {}
+    with open(keen_metadata, "r") as f:
+        for line in f:
+            line = json.loads(line)
+            if item_id_key in line and item_title_key in line:
+                iid2title[line[item_id_key]] = line[item_title_key][1:-1]
+    return iid2title
+
+
+def map_item_ids_to_sequential_ids(samples):
+    """
+    :param samples: dict of <user_id>: [<item_id>, <item_id>, ...]
+    :return:
+    """
+    uid2id, iid2id = {}, {}
+    for uid, ints in samples.items():
+        if uid not in uid2id:
+            uid2id[uid] = len(uid2id)
+        for iid in ints:
+            if iid not in iid2id:
+                iid2id[iid] = len(iid2id)
+
+    return uid2id, iid2id
+
+
+def create_splits(samples, do_random=False):
```

Comment:
Minor safety concern:
This will crash if someone decides to set the min_interactions parameter to 1 (to keep everything).

Reply:
Done.
Now you can also keep the shorter samples only for training, though they will not be evaluated